Don't extend from the deprecated Controller class

javiereguiluz opened the pull request:
This fixes the following deprecation:

```
The "Symfony\Bundle\FrameworkBundle\Controller\Controller" class is deprecated
since Symfony 4.2, use AbstractController instead.
```

But it creates an even more annoying deprecation:

```
Auto-injection of the container for "EasyCorp\Bundle\EasyAdminBundle\Controller\AdminController"
is deprecated since Symfony 4.2. Configure it as a service instead.
```

I don't know how can I solve that. I need your help. Thanks!

alterphp commented:
What if you make AdminController to implement ContainerAwareInterface ?

javiereguiluz commented:
Thank you all for your help. I tried defining the following service but the deprecation didn't disappeared. I'm probably not doing it right 😓 

```xml
        <service id="easyadmin.controller" class="EasyCorp\Bundle\EasyAdminBundle\Controller\AdminController" public="true">
            <tag name="container.service_subscriber" />
        </service>
```

yceruto commented:
@javiereguiluz I think the missing service is related to the controller that you're testing as well, so the tagged service `container.service_subscriber` should also be in `tests/Fixtures/App/config/config_override_controller.yml` for `OverridingEasyAdminController`.

alterphp commented:
IMO, making AdminController a service is a solution but it requires to adapt the routing configuration too.

See https://symfony.com/doc/current/controller/service.html#referencing-your-service-from-routing

javiereguiluz commented:
Thanks for your help. So, I did this:

* Removed the service for the controller
* Made the controller implement ContainerAwareInterface
* That's incompatible with AbstractController, so I stopped extending from it
* I copied the methods from AbstractController so it's like we're using it even if we are not actually using it

All deprecations are gone now ... although the resulting code is really ugly. I don't like it at all.

One test is failing though. The error message says:

```
The "property_accessor" service or alias has been removed or inlined when the container was compiled. You should either make it public, or stop using the container directly and use dependency injection instead.
```

I added `property_accessor` explicitly to the `getSubscribedServices()` list. Is this a bug of Symfony ... or am I doing something wrong? Thanks!

yceruto commented:
I'd suggest going back to the previous approach with `AbtractController` but fixing the configuration instead. For example something like this _should_ work:
```yaml
# config.yml
services:
    _defaults:
        autowire: true
        autoconfigure: true

    logger:
        class: Psr\Log\NullLogger

    EasyCorp\Bundle\EasyAdminBundle\Configuration\ConfigManager: '@easyadmin.config.manager'
    EasyCorp\Bundle\EasyAdminBundle\Search\QueryBuilder: '@easyadmin.query_builder'
    EasyCorp\Bundle\EasyAdminBundle\Search\Paginator: '@easyadmin.paginator'
    EasyCorp\Bundle\EasyAdminBundle\Search\Autocomplete: '@easyadmin.autocomplete'

    EasyCorp\Bundle\EasyAdminBundle\Tests\Fixtures\AppTestBundle\DataFixtures\AppFixtures:
        tags: ['doctrine.fixture.orm']

    EasyCorp\Bundle\EasyAdminBundle\Tests\Fixtures\AppTestBundle\Controller\:
        resource: '../../AppTestBundle/Controller'

    EasyCorp\Bundle\EasyAdminBundle\Tests\Fixtures\AppTestBundle\Admin\:
        resource: '../../AppTestBundle/Admin'

    EasyCorp\Bundle\EasyAdminBundle\Controller\AdminController: ~
```
You might need a little refactor to avoid duplicating this config for 
 `tests/Fixtures/App/config/config_customized_backend.yml` e.g. creates a separated file named `services.yml` and import it in `config.yml` and `config_customized_backend.yml`.

alterphp commented:
It makes sense to switch to controller as a service as it is the most durable implementation. And 2.0 release is the good time to BC Break !

javiereguiluz commented:
I'm not sure I can fix this myself 😅  Also, I'd prefer to not switch to a controller as service. I want to minimize the breaking changes ... even if this is a major version. So, if someone is up for the task, I'd appreciate if you could send a pull request to replace this one.

I want to release the EasyAdmin 2.0.0 stable version in 3 days (during the Hackday of SymfonyCon Lisbon 2018). If the fix for this "problem" is not ready, I don't think it's a big deal. This "error" is just a deprecation message. It doesn't break anything important. Thanks!

odolbeau commented:
Even is it's "only" a deprecated notice, it looks like the correction will result in a BC break anyway. As there is one planned in a few days, I think making the controller as service right now is a good idea.

javiereguiluz commented:
OK, if you think the service is the best idea ... I trust you!

yceruto commented:
Yes, extending it from `AbstractController` will break the access to custom services, but that's a good plan anyway for this release. See continuation here https://github.com/EasyCorp/EasyAdminBundle/pull/2490

alterphp commented:
The BC break I expect is about the way developers will embed the bundle (for now we have to define the routing from FQCN controller). And the way to controller will probably be modified.

But I think it's worth it anyway !

javiereguiluz commented:
Closing in favor of #2490.
